apps/prairielearn/assets/scripts/instructorAssessmentQuestionsClient.ts

```diff
@@ -1,8 +1,78 @@
+import { EditZoneModal } from '../../src/pages/instructorAssessmentQuestions/editZoneModal.html.js';
+import {
+  AssessmentQuestionRow,
+  AssessmentQuestionZone,
+} from '../../src/pages/instructorAssessmentQuestions/instructorAssessmentQuestions.types.js';
+
+import { EditAssessmentQuestionsTable } from './lib/editAssessmentQuestionsTable.js';
 import { histmini } from './lib/histmini.js';
 
 onDocumentReady(() => {
+  $('[data-toggle="popover"]').popover({ sanitize: false });
+
+  $('.js-sync-popover[data-toggle="popover"]')
+    .popover({ sanitize: false })
+    .on('show.bs.popover', function () {
+      $($(this).data('bs.popover').getTipElement()).css('max-width', '80%');
+    });
+
+  document.querySelectorAll<HTMLElement>('.js-histmini').forEach((element) => histmini(element));
+
+  const enableEditButton = document.getElementById('enableEditButton');
+  const editModeButtons = document.getElementById('editModeButtons');
+  const assessmentQuestionsTable = document.querySelector('.js-assessment-questions-table');
+  const assessmentType = (assessmentQuestionsTable as HTMLElement)?.dataset.assessmentType ?? '';
+  const urlPrefix = (assessmentQuestionsTable as HTMLElement)?.dataset.urlPrefix ?? '';
+  const assessmentInstanceId =
+    (assessmentQuestionsTable as HTMLElement)?.dataset.assessmentInstanceId ?? '';
+  const questions = decodeData('assessment-questions-data');
+
+  const zones: any = [];
+  let showAdvanceScorePercCol = false;
+  questions.forEach((question: AssessmentQuestionRow) => {
+    if (question.assessment_question_advance_score_perc !== 0) showAdvanceScorePercCol = true;
+    if (question.start_new_zone) {
+      const newZone = {
+        title: question.zone_title,
+        bestQuestions: question.zone_best_questions,
+        maxPoints: question.zone_max_points,
+        numberChoose: question.zone_number_choose,
+        questions: [],
+      };
+      zones.push(newZone);
+    }
+    if (question.alternative_group_size > 1) {
+      if (question.number_in_alternative_group === 1) {
+        const newAlternativeGroup = {
+          alternative_group_number: question.alternative_group_number,
+          alternative_group_number_choose: question.alternative_group_number_choose,
+          is_alternative_group: true,
+          init_points: question.init_points,
+          points_list: question.points_list,
+          max_points: question.max_points,
+          max_auto_points: question.max_auto_points,
+          max_manual_points: question.max_manual_points,
+          alternatives: [],
+        };
+        zones[zones.length - 1].questions.push(newAlternativeGroup);
+      }
+      question.is_alternative_group = true;
+      zones[zones.length - 1].questions[
+        zones[zones.length - 1].questions.length - 1
+      ].alternatives.push(question);
+    } else {
+      question.is_alternative_group = false;
+      zones[zones.length - 1].questions.push(question);
+    }
+  });
```

Comment:
This is reorganizing the data in a way that was easier to manipulate while editing and looks more similar to the format in the JSON files. After this part the data will look something like: 

```
zones: [{
    title: foo,
    bestQuestions,
    maxPoints,
    numberChoose:,
    questions: [
        {question data...}
     ],
 }]
 ```
 This is in contrast to the way the data comes directly is just an array of question objects with attributes like `zone_title` and `alternative_group_size` in order to format and put them in the table.
 
 However, I am not sure if this is the best way to do it. Doing it this way just made it much easier to comprehend the objects and manipulate them when writing some of the code below. I am open to discussing keeping this way or not formatting this way and manipulating the existing question objects in place, it just creates more code as there are more attributes to update for each change. The other possibility here would be to update the query to give us the data in a format like above without needing to manipulate it. This would require some rework of the existing table that displays the data without being in 'edit mode'. I will discuss this further in another comment where I talk about having a second table.

Reply:
I think manipulating the data as a "tree", just like it is in the underlying JSON, is the way to go. The only reason we use `zone_title`, `start_new_zone`, is because it's a little easier to write an SQL query that returns that instead of writing one that returns the whole tree structure.